File: leetcode/recursion/reverseList.java
**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseList(ListNode head) {
        // if [] return []
        if (head == null) {
            return head;
        
        // if [1] return [1]
        } else if (head.next == null) {
            return head;
        
        // else recursive case
        } else {
            // 1 -> 2 <- 3 <- new_head
            // 2 points to null also
            // wishful thinking: assume the tail has been solved
            ListNode new_head = reverseList(head.next);
            
            // 1 <=> 2 <- 3 <- new_head
            head.next.next = head;
            
            // null <- 1 <- 2 <- 3 <- new_head
            head.next = null;
            
            // new_head -> 3 -> 2 -> 1
            return new_head;
        }
        
    }
}

File: leetcode/recursion/reverseList/reverseList.py
# https://leetcode.com/explore/learn/card/recursion-i/251/scenario-i-recurrence-relation/3233/

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return head
        elif not head.next:
            return head
        else:
            
            new_head = Solution().reverseList(head.next)
            
            head.next.next = head
            head.next = None
            
            return new_head
        